src/StreamJsonRpc.Shared/JsonRpc.cs

```diff
@@ -654,6 +661,38 @@ protected virtual Task<ReturnType> InvokeCoreAsync<ReturnType>(int? id, string t
                 }
             }
         }
+        
+        private static Dictionary<string, List<string>> GetMethodToAttributeMap(object target)
+        {
+            var methodToAttributeMap = new Dictionary<string, List<string>>();
+            if (target != null)
+            {
+                for (TypeInfo t = target.GetType().GetTypeInfo(); t != null; t = t.BaseType?.GetTypeInfo())
+                {
+                    var anyMethodWithAttribute = t.DeclaredMethods.Any(m => m.GetCustomAttribute(typeof(MethodNameAttribute)) != null);
+                    if (anyMethodWithAttribute)
+                    {
+                        foreach (MethodInfo method in t.DeclaredMethods)
+                        {
+                            var attribute = method.GetCustomAttribute(typeof(MethodNameAttribute)) as MethodNameAttribute;
+                            if (attribute != null)
+                            {
+                                if (!methodToAttributeMap.ContainsKey(method.Name))
+                                {
+                                    methodToAttributeMap.Add(method.Name, new List<string>());
+                                }
+                                if (!methodToAttributeMap[method.Name].Contains(attribute.Name))
+                                {
+                                    methodToAttributeMap[method.Name].Add(attribute.Name);
+                                }
+                            }
```

Comment:
I'm not sure how critical performance will be for this method. But given we seek to be an perf-invisible interop layer, this block could be rewritten to have minimal dictionary lookups and key hashing:

```csharp
List<string> names;
if (!methodToAttributeMap.TryGetValue(method.Name, out names))
{
  names = new List<string>();
  if (!names.Contains(attribute.Name, StringComparer.Ordinal))
  {
    methodToAttributeMap.Add(attribute.Name, names);
  }
}

names.Add(attribute.Name);
```

I also threw in use of `StringComparer.Ordinal` in the example to state that we intentionally want a case sensitive search (which I *believe* we do).